elfo-network/src/frame/buffers.rs

```diff
@@ -0,0 +1,74 @@
+pub(crate) const COMPRESSED_DATA_BUFFER_CAPACITY: usize = 64 * 1024;
+pub(crate) const DECOMPRESSED_DATA_BUFFER_CAPACITY: usize = 128 * 1024;
+
+const BUFFER_CHUNK_SIZE: usize = 512;
+
+pub(crate) struct ReadBuffer {
+    buffer: Vec<u8>,
+    position: usize,
+    filled: usize,
+}
+
+impl ReadBuffer {
+    pub(crate) fn with_capacity(capacity: usize) -> Self {
+        Self {
+            buffer: Vec::with_capacity(capacity),
+            position: 0,
+            filled: 0,
+        }
+    }
+
+    pub(crate) fn as_slice(&self) -> &[u8] {
+        &self.buffer[self.position..self.filled]
+    }
+
+    pub(crate) fn remaining(&mut self) -> &mut [u8] {
+        &mut self.buffer[self.filled..]
+    }
+
+    pub(crate) fn mark_filled(&mut self, count: usize) {
+        debug_assert!(self.filled + count <= self.buffer.len());
+        self.filled += count;
+    }
+
+    pub(crate) fn consume(&mut self, count: usize) {
+        debug_assert!(self.position + count <= self.filled);
+        self.position += count;
+    }
+
+    pub(crate) fn len(&self) -> usize {
+        self.filled - self.position
+    }
+
+    pub(crate) fn extend_to_contain(&mut self, new_length: usize) {
+        // We round up the number of requested bytes to be a multiple of the chunk size
+        // to avoid reading too few bytes per syscall.
+        let current_length = self.len();
+        debug_assert!(new_length > current_length);
```

Comment:
I don't understand this assertion. `new_length` can easily be `4` (returned by the decoder), but the buffer can contain a lot of already-used bytes.